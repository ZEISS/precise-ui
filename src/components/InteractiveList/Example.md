**Elementary**

Most basic usage of Interactive list requires to specify data and whether the list is open or closed. The list is placed in absolutely positioned wrapper.

```jsx
const { InteractiveList } = require('precise-ui');

<InteractiveList data={['Item 1', 'Item 2', 'Item 3', 'Item 4 has very long text inside. It will break in two lines. Item 4 has very long text inside. It will break in two lines. Item 4 has very long text inside. It will break in two lines.']} open/>
```

You can show a tick for the selected item by using the `showTick` options.

```jsx
const { InteractiveList } = require('precise-ui');

<InteractiveList data={['Item 1', 'Item 2']} showTick open/>
```

Condensed list with smaller spacings

```jsx
const { InteractiveList } = require('precise-ui');

<InteractiveList data={['Item 1', 'Item 2', 'Item 3', 'Item 4 has very long text inside. It will break in two lines. Item 4 has very long text inside. It will break in two lines. Item 4 has very long text inside. It will break in two lines.']} condensed open/>
```

It's possible to select multiple elements, by adding checkboxes

```jsx
const { InteractiveList } = require('precise-ui');
<div style={{width: '50%'}}>
  <InteractiveList data={['Item 1', 'Item 2', 'Item 3', 'Item 4 has very long text inside. It will break in two lines. Item 4 has very long text inside. It will break in two lines. Item 4 has very long text inside. It will break in two lines.']} multiple open/>
</div>
```

You can add dividers
```jsx
const { InteractiveList } = require('precise-ui');

const data = ['1', '2',
{
  key: '1',
  type: 'divider',
}, {
  key: '2',
  content: 'Test Header',
}];
<InteractiveList data={data} open/>
```

**Custom Wrapper**

It's also possible to override the default wrapper by providing custom wrapper, where one could possibly build custom styled wrapper for the list.

```jsx

const { InteractiveList } = require('precise-ui');

const CustomWrapper = function (props) {
  const { children, ...rest } = props;
  const style = {
    WebkitBoxShadow: '0px 2px 5px 0px rgba(0,0,0,0.75)',
    MozBoxShadow: '0px 2px 5px 0px rgba(0,0,0,0.75)',
    boxShadow: '0px 2px 5px 0px rgba(0,0,0,0.75)',
  }
  return <div {...rest} style={style}>{children}</div>;
};

<InteractiveList data={['Value 1', 'Value 2', 'Value 3']} open customWrapper={CustomWrapper} />
```

**Controlled Mode**

Here is an example of using a controlling version of the interactive list together with another component.

```jsx
const { InteractiveList, Button } = require('precise-ui');

const items = [
  "Anim qui laborum officia laborum occaecat eu deserunt et.",
  "Value 2",
  "Value 3",
  "Value 4",
  "Value 5",
  "Value 6",
  "Value 7",
  "Value 8",
  "Value 9",
  "Value 10",
  "Value 11",
  "Value 12",
];

const ListContainerStyle = {
  position: 'absolute',
  width: '250px',
  maxHeight: '300px',
  zIndex: '999',
}

const CustomWrapper = ({ children, ...rest }) => {
  const style = {
    background: '#fff',
    WebkitBoxShadow: '0px 2px 5px 0px rgba(0,0,0,0.75)',
    MozBoxShadow: '0px 2px 5px 0px rgba(0,0,0,0.75)',
    boxShadow: '0px 2px 5px 0px rgba(0,0,0,0.75)',
  }
  return <div {...rest} style={style}>{children}</div>;
};

class CustomComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false
    }
  }

  handleMouseDown(e) {
    const { open } = this.state;
    this.setState({open: !open});
    e.preventDefault();
  }

  render() {
    const { open } = this.state;
    return (
      <div>
        <Button onMouseDown={this.handleMouseDown.bind(this)}>{ open ? 'Close' : 'Open' }</Button>
        <div style={ListContainerStyle}>
          <InteractiveList data={items} open={open} customWrapper={CustomWrapper}/>
        </div>
      </div>
    );
  }
}

<CustomComponent />
```

**Content Presentation**

Using links in the interactive list is possible (e.g., for building up menus).

```jsx
const { InteractiveList } = require('precise-ui');
const data = [
  {
    key: 'firstLink',
    content: <a href="#">First Link</a>,
  },
  {
    key: 'secondLink',
    content: <div><a href="#">Second Link</a></div>,
  },
];

<InteractiveList data={data} open/>
```

**Disable selection for specific items**

To disable specific items pass an array of strings (matching the strings if plain strings are used or the `key` 
field if objects are used as `data`). This prevents those items from being selected by the user.

Here is an example for single-selection:

```jsx
const { InteractiveList } = require('precise-ui');

const data = [
  'Value 1',
  'Value 2',
  'Value 3',
  {
    key: "4",
    content: "Value 4"
  },
];

const disabledItems = [
  'Value 2',
  '4',
];

<InteractiveList data={data} disabledItems={disabledItems} onChange={e => console.log('was changed', e)} open/>
```

Here is an example for multi-selection with the `multiple` option set to true.
 
```jsx
const { InteractiveList } = require('precise-ui');

const data = [
  'Value 1',
  'Value 2',
  'Value 3',
  {
    key: "4",
    content: "Value 4"
  },
];

const disabledItems = [
  'Value 2',
  '4',
];

<InteractiveList data={data} disabledItems={disabledItems} onChange={e => console.log('was changed', e)} multiple open/>
```

