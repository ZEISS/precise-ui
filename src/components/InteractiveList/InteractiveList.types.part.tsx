import { StandardProps } from '../../common';

export interface InteractiveListItemObject {
  key: string;
  content?: React.ReactChild;
  searchText?: string;
  type?: 'item' | 'divider' | 'header';
}

export type InteractiveListItem = InteractiveListItemObject | string;

export interface InteractiveListChangeEvent {
  /**
   * The changed value if any.
   */
  value: Array<number>;
}

export interface InteractiveListProps extends StandardProps {
  /**
   * The data to be displayed as selection basis.
   */
  data: Array<string | InteractiveListItem | undefined>;
  /**
   * Determines if the list should be open or not.
   */
  open?: boolean;
  /**
   * Determines if interactivity is disabled, i.e., touch etc. behaves as a scroll.
   */
  disabled?: boolean;
  /**
   * This Boolean attribute indicates that multiple options can be selected in the list.
   * If it is not specified, then only one option can be selected at a time.
   * Default value is `false`.
   */
  multiple?: boolean;
  /**
   * Optionally makes the accordion paddings smaller.
   */
  condensed?: boolean;
  /**
   * Switches the default to render a border on each list item.
   * Default value is `false`.
   */
  borderless?: boolean;
  /**
   * Removes the tick that is shown for the selected list item.
   * Default value is `false`.
   */
  showTick?: boolean;
  /**
   * The currently selected value, either a single one or multiple.
   * If used will go into controlled mode.
   */
  value?: string | Array<string>;
  /**
   * The initially selected value, either a single one or multiple.
   * Should be used to remain in managed mode.
   */
  defaultValue?: string | Array<string>;
  /**
   * Event triggered once the selected value changes.
   */
  onChange?(e: InteractiveListChangeEvent): void;
  /**
   * Event triggered once interactive list loses the focus.
   */
  onBlur?(): void;
  /**
   * Event triggered when a key was pressed.
   */
  onKeyDown?(e: React.KeyboardEvent<HTMLElement>): void;
  /**
   * The custom wrapper for rendering the list.
   */
  customWrapper?: React.ComponentType<InteractiveListWrapperProps>;
  /**
   * The content to consider for the interactive list.
   */
  children?: React.ReactNode;
  /**
   * @ignore
   */
  indices?: Array<number>;
  /**
   * Whether to position the list automatically based on screen size.
   */
  autoPosition?: boolean;
  /**
   * Whether to autofocus the list when it opens.
   */
  autoFocus?: boolean;
  /**
   * Set to true to focus the list.
   */
  focus?: boolean;
  /**
   * Set to false for the list to be visible as a shown block element.
   * Default value is `true`.
   */
  flyout?: boolean;
  /**
   * Event transported from the wrapper in case of a click somewhere inside.
   */
  onClick?(e: React.MouseEvent<HTMLElement>): void;
  /**
   * Event fired when clicked outside of the component
   */
  onClickOutside?(): void;
  /**
   * Whether to position the list to top or bottom.
   * Either `0` (bottom) or `1` (top). Make sure that the value provided is number.
   */
  direction?: InteractiveListDirection;

  /**
   * Disable specific items. Strings in this list have to match the strings used as `data` or the `key`
   * fields if `InteractiveListItemObject` are used.
   */
  disabledItems?: Array<string>;
}

export interface InteractiveListState {
  value: Array<number>;
  controlled: boolean;
  selected: number | undefined;
  direction: InteractiveListDirection;
}

export const enum InteractiveListBorderType {
  none = 0,
  normal = 1,
  focus = 2,
  error = 3,
}

export enum InteractiveListDirection {
  normal = 0,
  reverse = 1,
}

export interface InteractiveListWrapperProps {
  open: boolean;
  border: InteractiveListBorderType;
  direction: InteractiveListDirection;
  onClick?(e: React.MouseEvent<HTMLElement>): void;
}
