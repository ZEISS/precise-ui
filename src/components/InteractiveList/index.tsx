import * as React from 'react';
import onClickOutside, { AdditionalProps } from 'react-onclickoutside';
import { InteractiveListInt } from './InteractiveListInt.part';
import { InteractiveListProps } from './InteractiveList.types.part';

export {
  InteractiveListWrapperProps,
  InteractiveListDirection,
  InteractiveListBorderType,
  InteractiveListItem,
  InteractiveListProps,
  InteractiveListState,
  InteractiveListChangeEvent,
  InteractiveListItemObject,
} from './InteractiveList.types.part';

/**
 * General purpose list component which can be used for rendering list items.
 */
export const InteractiveList: React.ComponentClass<InteractiveListProps & AdditionalProps> = onClickOutside(
  InteractiveListInt,
);
InteractiveList.displayName = 'InteractiveList';
