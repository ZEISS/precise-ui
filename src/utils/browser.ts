const ua = (navigator && navigator.userAgent) || '';

function detectIE() {
  const msie = ua.indexOf('MSIE ');

  if (msie < 0) {
    const trident = ua.indexOf('Trident/');

    if (trident < 0) {
      const edge = ua.indexOf('Edge/');

      if (edge < 0) {
        return false;
      }

      return 'Edge';
    }

    return 'IE11';
  }

  return 'IE10';
}

export const isIE = detectIE();
