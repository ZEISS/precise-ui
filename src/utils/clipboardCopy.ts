export function clipboardCopy(content: string) {
  const el = document.createElement('textarea');
  el.value = content;
  el.setAttribute('readonly', '');
  el.style.position = 'absolute';
  el.style.left = '-9999px';
  document.body.appendChild(el);
  const selection = document && document.getSelection();

  if (selection) {
    const selected = selection.rangeCount > 0 ? selection.getRangeAt(0) : false;
    el.select();
    document.execCommand('copy');
    document.body.removeChild(el);

    if (selected) {
      selection.removeAllRanges();
      selection.addRange(selected);
    }
  }
}
