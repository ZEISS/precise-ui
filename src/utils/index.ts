export * from './browser';
export * from './clipboardCopy';
export * from './unit';
export * from './remCalc';
export * from './debounce';
export * from './stripUnit';
export * from './multiply';
export * from './colors';
export * from './interactivity';
export * from './colorCycle';
export * from './eventManager';
export * from './displayTo';
export * from './keyCodes';
export * from './text';
export * from './labels';
export {
  default as styled,
  css,
  keyframes,
  createGlobalStyle,
  withTheme,
  ThemeProvider,
  ThemeConsumer,
  ThemeContext,
  isStyledComponent,
  ServerStyleSheet,
  StyleSheetManager,
  themed,
} from './styled';
