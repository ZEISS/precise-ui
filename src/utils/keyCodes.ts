export const enum KeyCodes {
  enter = 13,
  space = 32,
  escape = 27,
  backspace = 8,
  delete = 46,
  end = 35,
  home = 36,
  left = 37,
  up = 38,
  right = 39,
  down = 40,
  comma = 188,
  semicolon = 186,
}
