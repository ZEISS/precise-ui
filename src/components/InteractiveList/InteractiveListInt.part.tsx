import * as React from 'react';
import styled, { themed, css } from '../../utils/styled';
import {
  InteractiveListWrapperProps,
  InteractiveListDirection,
  InteractiveListBorderType,
  InteractiveListItem,
  InteractiveListProps,
  InteractiveListState,
} from './InteractiveList.types.part';
import { Icon } from '../Icon';
import { Checkbox } from '../Checkbox';
import { KeyCodes } from '../../utils/keyCodes';
import { pacificBlue, transparent } from '../../colors';
import { distance } from '../../distance';
import { getFontStyle } from '../../textStyles';

interface ListItemProps {
  selected: boolean;
  hovered: boolean;
  disabled?: boolean;
  onClick?(e: React.MouseEvent): void;
  onMouseMove?(e: React.MouseEvent): void;
}

interface ListItemContentProps {
  condensed: boolean;
  showTick: boolean;
}

interface ListItemIconProps {
  condensed: boolean;
}

const InteractiveListContainer = styled.div`
  position: relative;
  outline: none;
`;

const ListWrapper = styled('ul')<InteractiveListWrapperProps>(
  themed(
    ({ open, border, direction, theme: { ui4 } }) => css`
      display: ${open ? 'block' : 'none'};
      list-style: none;
      width: 100%;
      position: relative;
      transform: translateY(${direction === InteractiveListDirection.normal ? 0 : -100}%);
      box-sizing: border-box;
      box-shadow: none;
      margin: 0;
      padding: 0;
      background: ${transparent};
      border: 1px solid ${border === InteractiveListBorderType.none ? transparent : ui4};
      ${direction === InteractiveListDirection.normal ? 'border-top: none' : 'border-bottom: none'};
      max-height: 100%;
      overflow-y: auto;
      z-index: 100;
    `,
  ),
);

const ListItem = styled.li<ListItemProps>(
  themed(
    ({ hovered, theme: { ui3, text2 }, disabled }) => css`
      ${getFontStyle({ size: 'medium' })}
      
      background: ${hovered ? ui3 : transparent};
      color: ${text2};
      list-style: none;
      box-sizing: border-box;
      opacity: ${disabled ? '0.5' : '1.0'};
      cursor: ${disabled ? 'not-allowed' : 'pointer'};
      display: block;
      width: 100%;
      height: auto;
      position: relative;

      a {
        color: inherit;
        display: block;
        text-decoration: none;

        &:hover {
          text-decoration: none;
        }
      }
    `,
  ),
);

const ListItemInnerContainer = styled.div`
  display: flex;
  align-items: center;
`;

const ListItemContent = styled.div`
  width: 100%;
`;

const ListItemContentPadding = styled('div')<ListItemContentProps>`
  padding: ${props => (props.condensed ? `${distance.small} ${distance.medium}` : distance.medium)};
  ${props => (props.showTick ? 'padding-right: 0;' : '')};
`;

const ListItemContentComponentPadding = styled('div')<ListItemContentProps>`
  > * {
    padding: ${props => (props.condensed ? `${distance.small} ${distance.medium}` : distance.medium)};
    ${props => (props.showTick ? 'padding-right: 0;' : '')};
  }
`;

const StyledCheckbox = styled(Checkbox)`
  display: block;
`;

const ListItemIconPadding = styled('div')<ListItemIconProps>`
  padding: 0 ${distance.medium};

  i {
    display: block;
  }
`;

const ListDivider = styled.hr`
  margin: ${distance.xsmall} ${distance.small};
  border: none;
  border-top: 1px solid ${themed(props => props.theme.textDisabled)};
`;

const ListHeader = styled.span`
  ${getFontStyle({ size: 'medium' })}
  
  padding: ${distance.medium} ${distance.small};
  width: 100%;
  display: block;
  height: auto;
  min-height: 1em;
  box-sizing: border-box;
  color: ${themed(props => props.theme.primary)};
`;

function getChosen(selected: string | Array<string>, multiple?: boolean) {
  if (multiple) {
    return Array.isArray(selected) ? selected : [selected];
  } else {
    return [Array.isArray(selected) ? selected[0] : selected];
  }
}

function getIndices(
  data: Array<InteractiveListItem | undefined>,
  selected: string | Array<string>,
  multiple?: boolean,
) {
  const chosen = getChosen(selected, multiple);
  const indices: Array<number> = [];

  for (let i = 0; i < data.length; i++) {
    const item = data[i];

    if (!item) {
      continue;
    }

    const key = typeof item === 'string' ? item : item.key;

    if (chosen.indexOf(key) !== -1) {
      indices.push(i);
    }
  }

  if (!multiple && indices.length > 1) {
    indices.splice(1);
  }

  return indices;
}

function autoSelect(element: HTMLElement) {
  const child = element && element.querySelector('a');

  if (child) {
    child.click();
  }
}

export class InteractiveListInt extends React.PureComponent<InteractiveListProps, InteractiveListState> {
  private readonly selects: Array<() => void> = [];
  private readonly elements: Array<HTMLElement> = [];
  private interactiveList: HTMLElement | null;

  constructor(props: InteractiveListProps) {
    super(props);
    const value = props.value || props.defaultValue || [];
    this.state = {
      value: props.indices || getIndices(props.data || [], value, props.multiple),
      controlled: props.indices !== undefined || props.value !== undefined,
      selected: undefined,
      direction: props.direction || InteractiveListDirection.normal,
    };
  }

  private defaultHandleClickOutside = () => {
    const { open, onBlur } = this.props;

    if (open) {
      if (typeof onBlur === 'function') {
        onBlur();
      }

      this.setState({
        selected: undefined,
      });
    }
  };

  handleClickOutside = () => {
    if (this.props.open) {
      this.props.onClickOutside ? this.props.onClickOutside() : this.defaultHandleClickOutside();
    }
  };

  componentWillReceiveProps(nextProps: InteractiveListProps) {
    const { open, autoPosition, focus } = this.props;

    if (this.state.controlled) {
      const { indices, value, data = [], multiple } = nextProps;

      this.setState({
        value: indices || getIndices(data, value || [], multiple),
      });
    }

    const componentIsNotFocused =
      nextProps.focus !== focus || (this.interactiveList && document.activeElement !== this.interactiveList);

    if (componentIsNotFocused && nextProps.focus) {
      if (open && nextProps.open) {
        this.interactiveList && this.interactiveList.focus();
        this.setState({
          selected: 0,
        });
      }
    }

    if (autoPosition && !open && nextProps.open && this.interactiveList) {
      const windowHeight = window.innerHeight;
      const { top } = this.interactiveList.getBoundingClientRect();

      if (top > windowHeight / 2) {
        this.setState(() => ({
          direction: InteractiveListDirection.reverse,
        }));
      } else {
        this.setState(() => ({
          direction: InteractiveListDirection.normal,
        }));
      }
    }
  }

  componentDidUpdate() {
    const { open, autoFocus } = this.props;

    if (open && autoFocus && this.interactiveList) {
      this.interactiveList.focus();
    }
  }

  private getNext(proposed: number, dir: number) {
    const { data } = this.props;
    const max = data.length;

    for (let i = 0; i < max; i++) {
      const item = data[proposed];

      if (!item) {
        continue;
      }

      if (typeof item === 'string' || (item.type !== 'divider' && item.type !== 'header')) {
        return proposed;
      }

      proposed += dir;

      if (proposed < 0) {
        proposed = max - 1;
      } else if (proposed === max) {
        proposed = 0;
      }
    }

    return 0;
  }

  private jumpTo(part: string) {
    const { data } = this.props;
    const { selected } = this.state;
    const start = selected ? (selected === data.length - 1 ? 0 : selected) : 0;
    let index = start;

    for (let i = start; i < start + data.length; i++, index = i === data.length ? 0 : index + 1) {
      const item = data[index];
      if (!item) {
        continue;
      }
      const lookup = typeof item === 'string' ? item : item.searchText || item.content || item.key;

      if (typeof lookup === 'string' && lookup.substr(0, part.length).toLowerCase() === part && selected !== index) {
        this.setState({
          selected: index,
        });
        break;
      }
    }
  }

  private control = (e: React.KeyboardEvent<HTMLElement>) => {
    const { data, onBlur, disabled, onKeyDown } = this.props;

    if (!disabled) {
      const { selected } = this.state;

      if (typeof onKeyDown === 'function') {
        onKeyDown(e);
      }

      switch (e.keyCode) {
        case KeyCodes.down:
          this.setState({
            selected: this.getNext(selected === undefined ? 0 : (selected + 1) % data.length, 1),
          });
          break;
        case KeyCodes.up:
          this.setState({
            selected: this.getNext(((selected || data.length) - 1) % data.length, -1),
          });
          break;
        case KeyCodes.home:
          this.setState({
            selected: this.getNext(0, 1),
          });
          break;
        case KeyCodes.end:
          this.setState({
            selected: this.getNext(data.length - 1, -1),
          });
          break;
        case KeyCodes.enter:
        case KeyCodes.space:
          if (typeof selected === 'number') {
            autoSelect(this.elements[selected]);
          }

          this.done();
          break;
        case KeyCodes.escape:
          if (typeof onBlur === 'function') {
            onBlur();
          }

          break;
        default:
          this.jumpTo(e.key.toLowerCase());
          return;
      }
    }

    e.stopPropagation();
    e.preventDefault();
  };

  private getItemDetails = (item: InteractiveListItem) => {
    let key = '';
    let content: React.ReactChild;

    if (typeof item === 'string') {
      key = item;
      content = item;
    } else {
      key = item.key;
      content = item.content || item.key;
    }

    return {
      key,
      content,
    };
  };

  private createSingleItem = (item: InteractiveListItem, index: number) => {
    const { key, content } = this.getItemDetails(item);

    return this.createItem(key, index, content);
  };

  private createMultipleItem = (item: InteractiveListItem, index: number) => {
    const { disabledItems } = this.props;
    const { value } = this.state;
    const { key, content } = this.getItemDetails(item);

    const isItemDisabled = disabledItems && disabledItems.indexOf(key) !== -1;

    const newContent = (
      <StyledCheckbox value={value.indexOf(index) >= 0} disabled={isItemDisabled}>
        {content}
      </StyledCheckbox>
    );
    return this.createItem(key, index, newContent);
  };

  private done = () => {
    const { multiple, onChange, disabled, data, disabledItems = [] } = this.props;
    const { value, controlled, selected } = this.state;

    const item = data && selected !== undefined && data[selected];
    const key = item && this.getItemDetails(item).key;
    const isItemDisabled = key && disabledItems.indexOf(key) !== -1;

    if (!disabled && !isItemDisabled && selected !== undefined) {
      const newValue = multiple
        ? value.indexOf(selected) === -1
          ? [...value, selected]
          : value.filter(i => i !== selected)
        : [selected];

      if (!controlled) {
        this.setState({
          value: newValue,
        });
      }

      if (typeof onChange === 'function') {
        onChange({
          value: newValue,
        });
      }
    }
  };

  private createItem(key: string, index: number, content: React.ReactChild) {
    const { value, selected } = this.state;
    const { disabled, showTick = false, multiple = false, condensed = false, disabledItems } = this.props;
    const isItemDisabled = disabledItems && disabledItems.indexOf(key) !== -1;
    const isHovered = !disabled && selected === index;
    const isSelected = value.indexOf(index) >= 0;
    const selects = this.selects;
    const ContentWrapper = typeof content === 'string' ? ListItemContentPadding : ListItemContentComponentPadding;

    if (selects[index] === undefined) {
      selects[index] = () => this.select(index);
    }

    return (
      <ListItem
        key={`${key}-${index}`}
        onClick={this.done}
        onMouseMove={selects[index]}
        selected={isSelected}
        hovered={isHovered}
        disabled={isItemDisabled}
        ref={(node: HTMLLIElement) => {
          this.elements[index] = node;

          if (isHovered && node) {
            const parent = node.parentElement;

            if (parent) {
              const height = parent.clientHeight;
              const position = parent.scrollTop;
              const y0 = node.offsetTop - position;
              const y1 = y0 + node.offsetHeight - height;

              if (y0 < 0) {
                parent.scrollTop = node.offsetTop;
              } else if (y1 > 0) {
                parent.scrollTop += y1;
              }
            }
          }
        }}>
        <ListItemInnerContainer>
          <ListItemContent>
            <ContentWrapper condensed={condensed} showTick={isSelected && showTick}>
              {content}
            </ContentWrapper>
          </ListItemContent>
          {isSelected && !multiple && showTick && (
            <ListItemIconPadding condensed={condensed}>
              <Icon name="Check" color={pacificBlue} size={1.375} />
            </ListItemIconPadding>
          )}
        </ListItemInnerContainer>
      </ListItem>
    );
  }

  private select(index: number) {
    const { disabled } = this.props;

    if (!disabled && this.state.selected !== index) {
      this.setState({
        selected: index,
      });
    }
  }

  private setNode = (ref: HTMLElement | null) => {
    this.interactiveList = ref;
  };

  render() {
    const {
      onChange: _0,
      indices: _1,
      disabled: _2,
      onKeyDown: _3,
      onClickOutside: _4,
      data = [],
      theme,
      borderless = false,
      children,
      multiple = false,
      open = false,
      onBlur,
      customWrapper,
      onClick,
      ...props
    } = this.props;
    const border = borderless ? InteractiveListBorderType.none : InteractiveListBorderType.normal;
    const createItem = multiple ? this.createMultipleItem : this.createSingleItem;
    const Wrapper = customWrapper || ListWrapper;

    return (
      <InteractiveListContainer
        ref={this.setNode}
        {...(open ? { tabIndex: 0 } : undefined)}
        onKeyDown={this.control}
        {...props}>
        <Wrapper open={open} border={border} direction={this.state.direction} onClick={onClick}>
          {open &&
            data.map((item, index) => {
              if (item) {
                if (typeof item !== 'string') {
                  switch (item.type) {
                    case 'divider':
                      return <ListDivider key={item.key} />;
                    case 'header':
                      return <ListHeader key={item.key}>{item.content || item.key}</ListHeader>;
                  }
                }

                return createItem(item, index);
              }

              return undefined;
            })}
        </Wrapper>
      </InteractiveListContainer>
    );
  }
}
