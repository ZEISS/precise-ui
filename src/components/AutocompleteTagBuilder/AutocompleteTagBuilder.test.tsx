import * as React from 'react';
import * as enzyme from 'enzyme';
import { AutocompleteTagBuilder } from './index';
import { InteractiveList } from '../InteractiveList';
import { KeyCodes } from '../../utils/keyCodes';
import 'jest-styled-components';

describe('<AutocompleteTagBuilder />', () => {
  it('should render <AutocompleteTagBuilder> component', () => {
    const wrapper = enzyme.shallow(
      <AutocompleteTagBuilder suggestions={['one', 'two', 'three']} defaultValue={['five']} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should show suggestions list when focused', () => {
    const wrapper = enzyme.mount(
      <AutocompleteTagBuilder suggestions={['one', 'two', 'three']} defaultValue={['five']} />,
    );
    // tslint:disable-next-line:no-string-literal
    global['cancelAnimationFrame'] = jest.fn();
    wrapper
      .find('div')
      .at(0)
      .prop('onFocus')({} as React.FocusEvent);
    expect(wrapper.update().find('li')).toHaveLength(3);

    // tslint:disable-next-line:no-string-literal
    delete global['cancelAnimationFrame'];
  });

  it('should maintain input focus after suggestion is selected', () => {
    const wrapper = enzyme.mount(
      <AutocompleteTagBuilder suggestions={['one', 'two', 'three']} defaultValue={['five']} />,
    );
    // tslint:disable-next-line:no-string-literal
    global['cancelAnimationFrame'] = jest.fn();
    wrapper
      .find('div')
      .at(0)
      .prop('onFocus')({} as React.FocusEvent);
    const suggestions = wrapper.update().find(InteractiveList);
    suggestions.simulate('keyDown', {
      keyCode: KeyCodes.down,
      stopPropagation() {},
      preventDefault() {},
    });
    suggestions.simulate('keyDown', {
      keyCode: KeyCodes.enter,
      stopPropagation() {},
      preventDefault() {},
    });

    expect(wrapper.find('input').getDOMNode()).toBe(document.activeElement);

    // tslint:disable-next-line:no-string-literal
    delete global['cancelAnimationFrame'];
  });
});
