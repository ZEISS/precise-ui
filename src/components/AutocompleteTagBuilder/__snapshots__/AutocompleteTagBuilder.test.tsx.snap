// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutocompleteTagBuilder /> should render <AutocompleteTagBuilder> component 1`] = `
<AutocompleteTagBuilderInt
  defaultValue={
    Array [
      "five",
    ]
  }
  suggestions={
    Array [
      "one",
      "two",
      "three",
    ]
  }
/>
`;
